javaslang/src/main/java/javaslang/collection/ArrayType.java

```diff
@@ -0,0 +1,147 @@
+            } else if (short.class == type) {
+                return (ArrayType<T>) ShortArrayType.INSTANCE;
+            } else {
+                throw new IllegalArgumentException("Unknown type: " + type);
+            }
+        }
+    }
+
+    Object newInstance(int length) { return copy(empty(), length); }
+
+    /** System.arrayCopy with same source and destination */
+    Object copyRange(Object array, int from, int to) {
+        final int length = to - from;
+        return copy(array, length, from, 0, length);
+    }
+
+    /** Repeatedly group an array into equal sized sub-trees */
+    Object grouped(Object array, int groupSize) {
+        final int arrayLength = lengthOf(array);
+        assert arrayLength > groupSize;
+        final Object results = obj().newInstance(1 + ((arrayLength - 1) / groupSize));
+        obj().setAt(results, 0, copyRange(array, 0, groupSize));
+
+        for (int start = groupSize, i = 1; start < arrayLength; i++) {
+            final int nextLength = Math.min(groupSize, arrayLength - (i * groupSize));
+            obj().setAt(results, i, copyRange(array, start, start + nextLength));
+            start += nextLength;
+        }
+
+        return results;
+    }
+
+    /** clone the source and set the value at the given position */
+    Object copyUpdate(Object array, int index, T element) {
+        final Object copy = copy(array, index + 1);
+        setAt(copy, index, element);
+        return copy;
+    }
+
+    Object copy(Object array, int minLength) {
+        final int arrayLength = lengthOf(array);
+        final int length = Math.max(arrayLength, minLength);
+        return copy(array, length, 0, 0, arrayLength);
+    }
+
+    /** clone the source and keep everything after the index (pre-padding the values with null) */
+    Object copyDrop(Object array, int index) {
+        final int length = lengthOf(array);
+        return copy(array, length, index, index, length - index);
+    }
+
+    /** clone the source and keep everything before and including the index */
+    Object copyTake(Object array, int lastIndex) {
+        return copyRange(array, 0, lastIndex + 1);
+    }
+
+    /** Create a single element array */
+    Object asArray(T element) {
+        final Object result = newInstance(1);
+        setAt(result, 0, element);
+        return result;
+    }
+
+    /** Store the content of an iterable in an array */
+    static Object[] asArray(java.util.Iterator<?> it, int length) {
+        final Object[] array = new Object[length];
+        for (int i = 0; i < length; i++) {
+            array[i] = it.next();
+        }
+        return array;
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T> T asPrimitives(Class<?> primitiveClass, Iterable<?> values) {
+        final Object[] array = Array.ofAll(values).toJavaArray();
+        assert (array.length == 0) || (primitiveClass == asPrimitive(array[0].getClass())) && !primitiveClass.isArray();
```

Comment:
If you still think I should, I will, just let me know if it's still the case (see my other responses)